### PR TITLE
[fx] remove delegate, replace with tracer

### PR DESCRIPTION
**Pull request by zdevito**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #44737 [fx] demo test that shows how to preserve module hierarchy
* **#44566 [fx] remove delegate, replace with tracer**

The Delegate objects were confusing. They were suppose to be a way to
configure how tracing works, but in some cases they appeared necessary
for consturcting graphs, which was not true. This makes the organization
clearer by removing Delgate and moving its functionality into a Tracer class,
similar to how pickle has a Pickler class.

Differential Revision: [D23683177](https://our.internmc.facebook.com/intern/diff/D23683177)

### CODE REVIEW

File: torch/fx/graph.py [hunk 1]
```diff
@@ -88,6 +88,31 @@ def create_node(self, op: str, target: Target,
...
 
+    # sugar for above when you know the op
+    def placeholder(self, name: str) -> Node:
+        return self.create_node('placeholder', name)
+
+    def get_param(self, name: str) -> Node:
+        return self.create_node('get_param', name)
+
+    def call_module(self, 
+                    module_name: str,
```
Comment (jamesr66a) [inline torch/fx/graph.py:99]: nit: Maybe name this `module_qualified_name` to emphasize that it's the full qualname?

Also probably a good idea to add docstrings for these methods

File: torch/fx/graph.py [hunk 2]
```diff
@@ -88,6 +88,31 @@ def create_node(self, op: str, target: Target,
...
+        return self.create_node('call_module', module_name, args, kwargs)
+
+    def call_method(self, 
+                    method_name: str, 
+                    args: Optional[Tuple[Argument, ...]] = None,
+                    kwargs: Optional[Dict[str, Argument]] = None) -> Node:
+        return self.create_node('call_method', method_name, args, kwargs)
+
+    def call_function(self, 
+                      the_function: Callable[..., Any],
```
Comment (jamesr66a) [inline torch/fx/graph.py:111]: Why `the_function` and not something like `function`?

File: torch/fx/graph.py [hunk 3]
```diff
@@ -115,6 +140,7 @@ def _name(self, target: Target) -> str:
             if _is_magic(op):
                 op = op[2:-2]
         op = op.replace('.', '_')
+        op = op.replace('*', '')
```
Comment (jamesr66a) [inline torch/fx/graph.py:143]: why?

File: torch/fx/proxy.py [hunk 4]
```diff
@@ -3,12 +3,57 @@
...
-from typing import TYPE_CHECKING, Tuple, Dict, Optional, Iterable, NoReturn, Any
-from .node import Target, Node
-
-if TYPE_CHECKING:
-    from .symbolic_trace import DelegateBase
+from .graph import magic_methods, reflectable_magic_methods, Graph
+from typing import Tuple, Dict, Optional, Iterable, NoReturn, Any, Union, Callable
+from .node import Target, Node, Argument, base_types
+
+class TracerBase:
```
Comment (jamesr66a) [inline torch/fx/proxy.py:10]: Kinda weird this is in `proxy.py`. Can it be in its own file?

**Review by jamesr66a — APPROVED** [review]
Looks fine, only nits

File: torch/fx/graph.py (hunk 3 above)
Comment (zdevito) [reply torch/fx/graph.py:143]: names need to be valid identifiers, but we were previously passing in '*args' to this function. I hit the bug refactoring around special handling for `get_param` operators which stripped this off, but only in the delegate.

File: torch/fx/proxy.py (hunk 4 above)
Comment (zdevito) [reply torch/fx/proxy.py:10]: I tried but it's cyclic: TracerBase depends on Proxy and Proxy depends on TracerBase.

File: torch/fx/graph.py (hunk 1 above)
Comment (zdevito) [reply torch/fx/graph.py:99]: ill rename it in the next in the stack since this is ready to land

**Pull request closed without merging**